[ConstraintSystem] Implement heuristics for pruning the generic operator overload search space

hborla opened the pull request:
* If the solver already found a solution with an operator disjunction choice that did not introduce any conversions, check conformance requirements of subsequent choices early using known argument types, and skip any choices with unsatisfied requirements.
* Within the generic operator disjunction partition, attempt the most specific choices first. If the solver finds a solution, it can then skip subsequent choices that can always be used in place of the successful choices. For example, consider the following overload set:
```swift
extension RangeReplaceableCollection {
  static func + <Other: RangeReplaceableCollection>(lhs: Self, rhs: Other) -> Self where Element == Other.Element
  static func + <Other: Sequence>(lhs: Other, rhs: Self) -> Self where Element == Other.Element
  static func + <Other: Sequence>(lhs: Self, rhs: Other) -> Self where Element == Other.Element
}
```

If the solver finds a solution with the first overload of `+`, it doesn't need to attempt the others because `RangeReplaceableCollection` refines `Sequence`, so the other two are guaranteed to work and produce the same solution, but the solver will never choose those overloads because they're less specific.

These two heuristics together are almost everything we need to remove the type variable merging hack from `LinkedExprAnalyzer`. The last step is making SIMD arithmetic code more efficient by changing the ordering of the disjunction partition to not try those overloads last when SIMD types are involved.

Resolves: rdar://65007946

hborla committed:
[ConstraintSystem] Always log disjunction choices that are skipped and
the reason for skipping while solving in debug mode.

hborla committed:
[ConstraintSystem] If the solver has already found a solution with a
disjunction choice that does not introduce conversions, check to see
if known argument types satisfy generic operator conformance requirements
early, and skip the overload choice if any requirements fail.

This helps the solver avoid exploring way too much search space when
the right solution involves a generic operator, but the argument types
are known up front, such as `collection + collection + collection`.

hborla committed:
[ConstraintSystem] Within the generic disjunction choice partition,
attempt the most specific choices first. Then, if the solver finds
a solution with one choice, it can skip any subsequent choices that
can be unconditionally used in place of the successful chioce and produce
the same solution.

hborla committed:
[ConstraintSystem] Requestify isDeclSubstitutable from CSStep.cpp

hborla committed:
[ConstraintSystem] Account for scoring in the new generic overload
heuristics.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Only attempt generic overload pruning heuristics for
operators for now.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

xedin commented on lib/Sema/CSStep.cpp:
```diff
@@ -525,29 +528,132 @@ StepResult DisjunctionStep::resume(bool prevFailed) {
+  // can be unconditionally substituted by the current choice, skip the current
+  // choice.
+  if (LastSolvedChoice && LastSolvedChoice->second == getCurrentScore() &&
+      choice.isGenericOperator()) {
```
I like the idea behind this, maybe it could be improved by doing this filtering only once upon encountering a new disjunction since `TypeChecker::checkGenericArguments` seems like an expensive operation to repeat and substitution always means global best?

hborla replied:
Yes, I'm working on making the solver check this less often, but the nice thing about this request is that it's independent of what's being type checked, so it can be cached and reused again across constraint systems. This is not true of `compareDeclarations`, so I am just not going to use it in these heuristics. Also, the substitution is guaranteed to produce the same function type so in theory it should not affect overload resolution elsewhere in the expression (though for this to be 100% true I think I need to change the request to make sure the argument and result types are plain type parameters)

xedin replied:
bq. it can be cached and reused again across constraint systems

That makes sense but my point is more of a logical place to do that check, I think it makes sense to filter on arrival since this is a stable operation even if it does cache the result.

hborla replied:
Ah I see, the issue is that we can only filter out more general overloads if the solver finds a solution with the more refined one.

A more sophisticated version of this that @DougGregor pointed out is we can identify overloads that are pure refinements of a more general overload, and make the solver only try the most general one. Then, the solver can select a more refined overload at the end once it has found a solution with the most general one. I think that's out of scope for this PR, but a super interesting idea that we should think through.

xedin commented on lib/Sema/CSStep.h:
```diff
@@ -652,6 +654,29 @@ class DisjunctionStep final : public BindingStep<DisjunctionChoiceProducer> {
+    if (!typeVar)
+      return;
+
+    auto result = cs.findConstraintThroughOptionals(
```
This is definitely would be great to remove before merging as per your FIXME because it adds one more usage of `gatherConstraints` which is expensive.

hborla replied:
✅

hborla commented on lib/Sema/CSStep.cpp:
```diff
@@ -525,29 +528,132 @@ StepResult DisjunctionStep::resume(bool prevFailed) {
+    auto signature = genericFnType->getGenericSignature();
+
+    for (auto argParamPair : llvm::zip(argFnType->getParams(),
+                                       choiceFnType->getParams())) {
```
This is only done for operator calls, which to my knowledge can't have default parameters...right?

xedin replied:
Sorry yeah, comment speaks about generic overloads so I didn't notice that it was isGeneric*Operator*() in the check...

hborla replied:
Whoops, I will fix up the comment, thanks! I had originally written some of this thinking it would work for all generic overloads, and then realized why it wouldn't haha

hborla commented on lib/Sema/CSStep.cpp:
```diff
@@ -525,29 +528,132 @@ StepResult DisjunctionStep::resume(bool prevFailed) {
+        continue;
+
+      for (auto *protocol : signature->getRequiredProtocols(paramType)) {
+        if (!TypeChecker::conformsToProtocol(argType, protocol, useDC))
```
This only checks conformance requirements directly on the type parameter that would be matched up with the argument, so it won't check any requirements that depend on other argument types

hborla committed:
[ConstraintSystem] Use isDeclSubstitutable instead of compareDeclarations
to order generic operator overload choices.

hborla committed:
[ConstraintSystem] Rename IsDeclSubstitutableRequest to
IsDeclRefinementOfRequest to better reflect what it computes.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Account for duplicate archetype bindings
in IsDeclRefinementOfRequest.

hborla committed:
[ConstraintSystem] Cache applied disjunction constraints in the
constraint system to use later in DisjunctionStep.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Only attempt the refinement overload heuristic
for arithmetic operators.

Only sort overloads that are related, e.g. Sequence
overloads. Further, choose which generic overloads
to attempt first based on whether any known argument types
conform to one of the standard arithmetic protocols.

hborla commented:
@swift-ci please test compiler performance

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Only do the work of partitioning the generic operator
overloads if generic operators are not going to be skipped.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Allow early conformance checking for overload choices
if the best solution has a non-default literal type.

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test compiler performance

hborla committed:
[DisjunctionStep] Only check requirements early to skip disjunction
choices for arithmetic operators for now.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

xedin commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -2177,6 +2177,11 @@ class ConstraintSystem {
+  /// A map from applied disjunction constraints to the corresponding
+  /// argument function type.
+  llvm::SmallMapVector<ConstraintLocator *, const FunctionType *, 4>
+      AppliedDisjunctions;
```
What do you think about the alternative idea here of storing pointer to argument function in the disjunction locator?

hborla replied:
Oh, that would be much better. I'm going to merge this now (the release source compat failure looks unrelated), and see if I can make that change as a follow on cleanup. My only thought is we might not have the argument function at the time when the disjunction constraint is created

xedin reviewed (approved):
LGTM! Left a minor suggestion inline.
